lib/internal/shutil.js

```diff
@@ -0,0 +1,294 @@
+      });
+  });
+}
+
+function fixWinEPERMSync(path, options, error) {
+  validatePath(path);
+  validateOptions(options);
+  validateError(error);
+
+  try {
+    options.chmodSync(path, _0666);
+  } catch (er2) {
+    if (er2.code === 'ENOENT') return;
+    else throw error;
+  }
+
+  try {
+    var stats = options.statSync(path);
+  } catch (er3) {
+    if (er3.code === 'ENOENT') return;
+    else throw error;
+  }
+
+  if (stats.isDirectory()) rmdirSync(path, options, error);
+  else options.unlinkSync(path);
+}
+
+function rmdir(path, options, originalEr, callback) {
+  validatePath(path);
+  validateOptions(options);
+  validateError(originalEr);
+  validateCallback(callback);
+
+  // try to rmdir first, and only readdir on ENOTEMPTY or EEXIST (SunOS)
+  // if we guessed wrong, and it's not a directory, then
+  // raise the original error.
+  options.rmdir(path, function(er) {
+    if (er && (er.code === 'ENOTEMPTY' || er.code === 'EEXIST' || er.code === 'EPERM'))
+      rmkids(path, options, callback);
+    else if (er && er.code === 'ENOTDIR') callback(originalEr);
+    else callback(er);
+  });
+}
+
+function rmdirSync(path, options, originalEr) {
+  validatePath(path);
+  validateOptions(options);
+  validateError(originalEr);
+
+  try {
+    options.rmdirSync(path);
+  } catch (er) {
+    if (er.code === 'ENOENT') return;
+    if (er.code === 'ENOTDIR') throw originalEr;
+    if (er.code === 'ENOTEMPTY' || er.code === 'EEXIST' || er.code === 'EPERM')
+      rmkidsSync(path, options);
+  }
+}
+
+function rmkids(path, options, callback) {
+  validatePath(path);
+  validateOptions(options);
+  validateCallback(callback);
+
+  options.readdir(path, function(er, files) {
+    if (er) return callback(er);
+    var n = files.length;
+    if (n === 0) return options.rmdir(path, callback);
+    var errState;
+    files.forEach(function(f) {
+      rmtree(join(path, f), options, function(er) {
+        if (errState) return;
+        if (er) return callback((errState = er));
+        if (--n === 0) options.rmdir(path, callback);
+      });
+    });
+  });
+}
+
+function rmkidsSync(path, options) {
+  validatePath(path);
+  validateOptions(options);
+
+  options.readdirSync(path).forEach(function(f) {
+    rmtreeSync(join(path, f), options);
+  });
+
+  // We only end up here once we got ENOTEMPTY at least once, and
+  // at this point, we are guaranteed to have removed all the kids.
+  // So, we know that it won't be ENOENT or ENOTDIR or anything else.
+  // try really hard to delete stuff on windows, because it has a
+  // PROFOUNDLY annoying habit of not closing handles promptly when
+  // files are deleted, resulting in spurious ENOTEMPTY errors.
+  var retries = isWindows ? 100 : 1;
+  var i = 0;
+  do {
+    var threw = true;
+    try {
+      var ret = options.rmdirSync(path, options);
+      threw = false;
+      return ret;
+    } finally {
+      if (++i < retries && threw) continue;
+    }
+  } while (true);
+}
+
+module.exports = {
```

Comment:
I think one thing we mentioned was that we should have a `Promise`-based implementation.  Since this is JS, maybe we can get away with consuming `util.promisify()`?

What I *don't* recall is whether or not this PR should block on the `Promise` implementation.  But it doesn't seem (?) so difficult to get it in.